test/k8s/res_limits_test.go

```diff
@@ -0,0 +1,485 @@
+			insNum, _ := os.LookupEnv("PERF_TEST_INSTANCES_PER_SERVICE")
+
+			if svcNum != "" {
+				numServices, _ = strconv.Atoi(svcNum)
+			}
+			if insNum != "" {
+				instancesPerService, _ = strconv.Atoi(insNum)
+			}
+		})
+
+		BeforeEach(func() {
+			Expect(cluster.Install(NamespaceWithSidecarInjection(TestNamespace))).To(Succeed())
+		})
+
+		AfterEach(func() {
+			Expect(cluster.DeleteNamespace(TestNamespace)).To(Succeed())
+		})
+
+		// should not crash, dump memory usage
+		It("should deploy all services and instances", func() {
+
+			adjustResource(cpu, memory, false)
+
+			deployDPs()
+
+			errCh := make(chan error)
+			ctx, cancelMonitoring := context.WithCancel(context.Background())
+			defer cancelMonitoring()
+			watchControlPlane(ctx, errCh)
+
+			dpCh := make(chan bool)
+			waitForDPs(dpCh)
+
+			var err error
+			select {
+			case dpRet := <-dpCh:
+				Logf("dpCh returned\n")
+				Expect(dpRet).To(BeTrue(), "data planes should be run and available")
+			case err = <-errCh:
+				Logf("errCh returned\n")
+			}
+			Expect(err).ToNot(HaveOccurred(), "control plane should not crash")
+		})
+
+		// half the resource, should crash without GOMEMLIMIT
+		It("should crash when deploy all services and instances with half CP resource", func() {
+			adjustResource(cpu/2, memory/2, false)
+
+			deployDPs()
+
+			errCh := make(chan error)
+			ctx, cancelMonitoring := context.WithCancel(context.Background())
+			watchControlPlane(ctx, errCh)
+
+			dpCh := make(chan bool)
+			waitForDPs(dpCh)
+
+			var err error
+			select {
+			case <-dpCh:
+				Logf("dpCh returned\n")
+			case err = <-errCh:
+				Logf("errCh returned\n")
+			}
+			cancelMonitoring()
+			Expect(err).To(HaveOccurred(), "control plane should crash with half resource")
+
+			cpPods, err := k8s.ListPodsE(cluster.GetTesting(), cluster.GetKubectlOptions(Config.KumaNamespace), metav1.ListOptions{
+				LabelSelector: fmt.Sprintf("app=%s", Config.KumaServiceName),
+			})
+			Expect(err).ToNot(HaveOccurred())
+			for _, pod := range cpPods {
+				if k8s.IsPodAvailable(&pod) {
+					continue
+				}
+
+				_, _ = fmt.Fprintf(GinkgoWriter, "Pod %s: %s\n", pod.Name, pod.Status.Phase)
+				for _, cts := range pod.Status.ContainerStatuses {
+					_, _ = fmt.Fprintf(GinkgoWriter, "Pod %s - %s: %s\n", pod.Name, cts.Name, cts.State.String())
+				}
+			}
+		})
+
+		It("should not crash when control plane has GOMEMLIMIT set when all services and instances with half CP resource", func() {
+			adjustResource(cpu/2, memory/2, true)
+
+			deployDPs()
+
+			errCh := make(chan error)
+			ctx, cancelMonitoring := context.WithCancel(context.Background())
+			watchControlPlane(ctx, errCh)
+
+			dpCh := make(chan bool)
+			waitForDPs(dpCh)
+
+			var err error
+			select {
+			case <-dpCh:
+				Logf("dpCh returned\n")
+			case err = <-errCh:
+				Logf("errCh returned\n")
+			}
+			cancelMonitoring()
+			Expect(err).ToNot(HaveOccurred(), "control plane should not crash")
+			if err != nil {
```

Comment:
There is no way we'll get there. Ginkgo will stop execution of this `It(` when err is not nil on this line `Expect(err).ToNot(HaveOccurred(), "control plane should not crash")`